EventHandlers/PrebidMobileAdMobAdapters/PrebidAdMobBannerAdapter.swift

```diff
@@ -0,0 +1,124 @@
+/*   Copyright 2018-2021 Prebid.org, Inc.
+ 
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+ 
+  http://www.apache.org/licenses/LICENSE-2.0
+ 
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  */
+
+import Foundation
+import PrebidMobile
+import GoogleMobileAds
+
+@objc(PrebidAdMobBannerAdapter)
+public class PrebidAdMobBannerAdapter:
+    NSObject,
+    GADCustomEventBanner,
+    DisplayViewLoadingDelegate,
+    DisplayViewInteractionDelegate {
+    
+    public var delegate: GADCustomEventBannerDelegate?
+    
+    public var displayView: PBMDisplayView?
+    
+    required public override init() {
+        super.init()
+    }
+    
+    
+    public func requestAd(_ adSize: GADAdSize, parameter serverParameter: String?, label serverLabel: String?, request: GADCustomEventRequest) {
+        
+        guard let eventExtras = request.additionalParameters, !eventExtras.isEmpty else {
+            let error = AdMobAdaptersError.emptyCustomEventExtras
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        guard let bid = eventExtras[PBMMediationAdUnitBidKey] as? Bid else {
+            let error = AdMobAdaptersError.noBidInEventExtras
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        guard let configId = eventExtras[PBMMediationConfigIdKey] as? String else {
+            let error = AdMobAdaptersError.noConfigIDInEventExtras
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        guard let serverParameter = serverParameter else {
+            let error = AdMobAdaptersError.noServerParameter
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        guard let jsonServerParameters = AdMobMediationHelper.stringToDictionaryString(dataString: serverParameter) else {
+            let error = AdMobAdaptersError.serverParameterInWrongFormat
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        guard let keywords = request.userKeywords as? [String] else {
+            let error = AdMobAdaptersError.emptyCustomEventExtras
+            delegate?.customEventBanner(self, didFailAd: error)
+            return
+        }
+        
+        for parameter in jsonServerParameters {
+            guard keywords.contains(parameter) else {
```

Comment:
We should introduce a more reliable way for matching the targeting keywords and the Ad Source parameters. Verification of the whole string makes as extremely format-sensitive. The single extra space inside the JSON structure will break the result. 
Need to compare string literals of the keys and values separately. And provide respective unit tests.
We should introduce a more reliable way for matching the targeting keywords and the Ad Source parameters. Verification of the whole string makes as extremely format-sensitive. The single extra space inside the JSON structure will break the result. 
Need to compare string literals of the keys and values separately. And provide respective unit tests.